### ISSUE TITLE
Add a way to programmatically set a gapcursor somewhere

### ISSUE THREAD

**Issue by carlobeltrame** [ProseMirror/prosemirror#1201]
### Issue details
I'm building a tiptap-based editor with some custom block elements. There are some controls that open a dialog and let the user choose an element to insert, after which the modal is closed, and the element is inserted at the current cursor position. In these cases, I need to place a (gap)cursor right after the inserted element.

Another use case would be on touch devices that have no arrow keys. It is hard or sometimes impossible to set the cursor between two block elements on these devices. If I could programmatically set the gapcursor at a position, I could do so when the user e.g. taps on a block element.

For now, I am working around this by patching the `handleClick` method on gapCursor to tolerate clicks outside of the editor (which result in `view.posAtCoords` returning `null`):
```js
    const gapCursorPlugin = gapCursor()
    // Patch the handleClick function so we can call it ourselves when autofocusing
    gapCursorPlugin.props.handleClick = (view, pos, event) => {
      if (!view.editable) return false
      let $pos = view.state.doc.resolve(pos)
      if (!GapCursor.valid($pos)) return false

      // The next two lines are patched to tolerate calculatedPos === null
      const calculatedPos = view.posAtCoords({left: event.clientX, top: event.clientY})
      if (calculatedPos && calculatedPos.inside > -1 && NodeSelection.isSelectable(view.state.doc.nodeAt(calculatedPos.inside))) return false

      view.dispatch(view.state.tr.setSelection(new GapCursor($pos)))
      return true
    }
    editor.registerPlugin(gapCursorPlugin)
```
Later, I can programmatically set the gapCursor as follows:
```js
gapCursorPlugin.props.handleClick(editor.view, myDesiredCursorPos, { clientX: -10000, clientY: -10000 })
```
This is obviously quite hacky, and I'd love to see a better way of acheiving the same result.

### Steps to reproduce

None so far, since this is more of a conceptual feature request. Let me know if I should provide an editor to play around with.

### ProseMirror version

prosemirror-gapcursor 1.1.5

### Affected platforms

- [x] Chrome
- [x] Firefox
- [?] Internet Explorer
- [x] Other: Especially touch devices

### Screencast

This is what my desired solution looks like, implemented using my workaround from above:

https://user-images.githubusercontent.com/7566995/132019446-051a3a66-d74b-4c81-84d6-bcaab272bda3.mp4

**Comment by marijnh** [comment]
You should be able to import `GapCursor` from `prosemirror-gapcursor`, create an instance with `new GapCursor($resolvedPos)`, and set the editor selection to that. Does that help?

**Comment by carlobeltrame** [comment]
Thanks for the input, that does help somewhat. I think I had this solution before, but needed a way to insert a normal cursor or a gapcursor, depending on what's more fitting at the resolved pos. Ideally, as an editor developer, I feel like I shouldn't have to think about which cursor is more appropriate, there is always one that is clearly better.
I gravitated towards my `handleClick` workaround because that method already makes this decision for me. So maybe my desire as a dev would be better stated as "Add a way to programmatically set the appropriate cursor somewhere".

For now, I improved my workaround using `if (GapCursor.valid(resolvedPos))`. But `handleClick` makes some more assertions I don't fully understand, based on `NodeSelection.isSelectable(view.state.doc.nodeAt(inside))`.
